### ISSUE TITLE
echo example not working / producing output on android

### ISSUE THREAD

**Issue by alhaad** [domokit/mojo#380]
When I run the echo example using the command:
```
mojo/devtools/common/mojo_run https://core.mojoapps.io/echo_client.mojo --android
```
no output is produced. Whereas it works perfectly for Linux.

I am facing a similar situation for a new service I am trying to add (sharing on android).
/cc @MitchRudominer @ppiet

**Comment by MitchRudominer** [comment]
It may just be the pulling of standard out  from the device that is not working. When I run the following command:

`mojo/devtools/common/mojo_run --android mojo:java_echo_client --logcat-tags=JavaEchoClient`

I see the output `I/JavaEchoClient(21671): Response: Hello Java!`

**Comment by MitchRudominer** [comment]
While the command that alhaad mentions in the first comment is running in one shell window I run in a second shell window `adb shell cat /data/data/org.chromium.mojo.shell/stdout.fifo` and I see no output. I think that the process is not writing standard out to the fifo.

**Comment by MitchRudominer** [comment]
By adding a logging statement I can confirm that the code
`base::android::RedirectStream(stdout, fifo_path, "w")` on line 179 of  `shell/android/main.cc` is being executed with the value of fifo_path equal to
`"/data/data/org.chromium.mojo.shell/stdout.fifo"`.

**Comment by MitchRudominer** [comment]
Not sure why but logging from echo_client running on Android successfully gets messages printed to the console but printf doesn't. See https://codereview.chromium.org/1291993003/.

**Comment by ppiet** [comment]
This is because printf prints to stdout and an app process on Android has its stdout redirected straight to /dev/null. LOG uses an Android NDK API to write to logcat.

**Comment by MitchRudominer** [comment]
Oh I thought that stdout on an app process was redirected to the fifo. Has this change recently or did echo client running on Android never print output on the console? After Alhaad filed this bug I was convinced into remembering that it used to work. But maybe this is a false memory.

**Comment by ppiet** [comment]
Oh, good point. I will take a look.

**Comment by ppiet** [comment]
I looked into it. The stdout redirection when running `mojo_run --android` seems to be working fine, however the stdout does not seem to be buffered by line. Ie. I had to add a fflush call after the printf to see the output.

Given that the shell process is not connected to the interactive terminal, this probably works as "intended" (by the std lib authors). I wonder if we should use setvbuf to enforce line buffering.
